src/Ledger/Deleg.lagda

```diff
@@ -133,41 +140,50 @@ data _⊢_⇀⦇_,POOL⦈_ : PoolEnv → PState → DCert → PState → Set whe
 
 data _⊢_⇀⦇_,VDEL⦈_ : VDelEnv → VState → DCert → VState → Set where
   VDEL-regdrep : let open PParams pp in
-    d ≡ poolDeposit -- TODO use drepDeposit instead
-    → c ∉ dReps
+    (d ≡ drepDeposit × c ∉ dom (dReps ˢ)) ⊎ (d ≡ 0 × c ∈ dom (dReps ˢ))
     ────────────────────────────────
-    pp ⊢ ⟦ dReps , ccKeys ⟧ᵛ ⇀⦇ regdrep c d an ,VDEL⦈
-         ⟦ ❴ c ❵ ∪ dReps , ccKeys ⟧ᵛ
+    ⟦ e , pp , vs ⟧ᶜ ⊢ ⟦ dReps , ccKeys ⟧ᵛ ⇀⦇ regdrep c d an ,VDEL⦈
+                       ⟦ ❴ c , e +ᵉ' drepActivity ❵ᵐ ∪ᵐˡ dReps , ccKeys ⟧ᵛ
 
   VDEL-deregdrep :
-    c ∉ dReps'
-    → ❴ c ❵ ∪ dReps' ≡ dReps
+    c ∈ dom (dReps ˢ)
     ────────────────────────────────
-    pp ⊢ ⟦ dReps , ccKeys ⟧ᵛ ⇀⦇ deregdrep c ,VDEL⦈
-         ⟦ dReps' , ccKeys ⟧ᵛ
+    Γ ⊢ ⟦ dReps , ccKeys ⟧ᵛ ⇀⦇ deregdrep c ,VDEL⦈
+        ⟦ dReps ∣ ❴ c ❵ ᶜ , ccKeys ⟧ᵛ
 
   VDEL-ccreghot :
     (kh , nothing) ∉ ccKeys ˢ
     -- TODO: Should we check if kh actually belongs to the CC?
     ────────────────────────────────
-    pp ⊢ ⟦ dReps , ccKeys ⟧ᵛ ⇀⦇ ccreghot (inj₁ kh) mkh ,VDEL⦈
-         ⟦ dReps , singletonᵐ kh mkh ∪ᵐˡ ccKeys ⟧ᵛ
+    Γ ⊢ ⟦ dReps , ccKeys ⟧ᵛ ⇀⦇ ccreghot (inj₁ kh) mkh ,VDEL⦈
+        ⟦ dReps , singletonᵐ kh mkh ∪ᵐˡ ccKeys ⟧ᵛ
 
 data _⊢_⇀⦇_,CERT⦈_ : CertEnv → CertState → DCert → CertState → Set where
   CERT-deleg :
     pp ⊢ stᵈ ⇀⦇ dCert ,DELEG⦈ stᵈ'
     ────────────────────────────────
-    pp ⊢ st ⇀⦇ dCert ,CERT⦈ record st { dState = stᵈ' }
+    ⟦ e , pp , vs ⟧ᶜ ⊢ st ⇀⦇ dCert ,CERT⦈ record st { dState = stᵈ' }
 
   CERT-vdel :
-    pp ⊢ stᵛ ⇀⦇ dCert ,VDEL⦈ stᵛ'
+    Γ ⊢ stᵛ ⇀⦇ dCert ,VDEL⦈ stᵛ'
     ────────────────────────────────
-    pp ⊢ st ⇀⦇ dCert ,CERT⦈ record st { vState = stᵛ' }
+    Γ ⊢ st ⇀⦇ dCert ,CERT⦈ record st { vState = stᵛ' }
 
   CERT-pool :
     pp ⊢ stᵖ ⇀⦇ dCert ,POOL⦈ stᵖ'
     ────────────────────────────────
-    pp ⊢ st ⇀⦇ dCert ,CERT⦈ record st { pState = stᵖ' }
+    ⟦ e , pp , vs ⟧ᶜ ⊢ st ⇀⦇ dCert ,CERT⦈ record st { pState = stᵖ' }
+
+data _⊢_⇀⦇_,CERTBASE⦈_ : CertEnv → CertState → ⊤ → CertState → Set where
```

Comment:
Where is this going to be used?

Reply:
Good point! This should be the base case of `CERT`, but the `SS⇒BS` transformer doesn't current support that.